cool.py

```diff
@@ -0,0 +1,294 @@
+        if not out_dir.exists():
+            out_dir.mkdir(parents=True)
+
+        mosaic_outfile = out_dir / f"{tile_name}.jpg"
+        logging.debug("saving to %s", mosaic_outfile)
+        cv2.imwrite(str(mosaic_outfile), mosaic_image)
+
+        return mosaic_image
+
+    else:
+        logging.debug("no output directory provided")
+
+        return mosaic_image
+
+
+def load_pytorch_model():
+    """load pytorch model with tower scout weights
+
+    Args:
+        None
+
+    Returns:
+        pytorch model: model ready for scanning
+    """
+    model_weight_path = Path(__file__).parent / 'tower_scout' / 'xl_250_best.pt'
+    yolov5_path = Path(__file__).parent / 'yolov5'
+    model = torch.hub.load(str(yolov5_path), 'custom', path=str(model_weight_path), source='local')
+    #: if statements (check that .pt file exists, folders, etc.)
+    
+    return model
+
+
+def get_model():
+    """gets pytorch model using logic to ensure it's only loaded once
+
+    Args:
+        None
+
+    Returns:
+        model: loaded pytorch model ready for use
+    """
+    global MODEL
+    if MODEL is None:
+        logging.info('loading pytorch model')
+        MODEL = load_pytorch_model()
+
+    return MODEL
+
+
+def detect_towers(image):
+    """run pytorch model with tower scout weight on an image to detect cooling towers
+
+    Args:
+        image (obj): Path object to local image file or np.ndarray object of in-memory file
+
+    Returns:
+        result (obj): pytorch result object
+    """
+    towerscout_model = get_model()
+        
+    # if isinstance(image, Path):
+    #     logging.info('working with a Path')
+    #     pil_image = Image.open(image)
+    # elif isinstance(image, np.ndarray):
+    #     logging.info('working with a numpy array')
+    #     pil_image = Image.fromarray(image)
+
+    logging.info('working with %s', type(image))
+
+    scan_start = perf_counter()
+    results = towerscout_model(image)
+    logging.info('time to scan image: %s',  format_time(perf_counter() - scan_start))
+
+    results.save()
+
+    return results
+
+
+def parse_results(results, col, row):
+    """parse detection results and calculate coordinates
+
+    Args:
+        results_df (dataframe): dataframe with results from pytorch model
+        col (str): the column of the WMTS index for the tile of interest (top-left tile)
+        row (str): the row of the WMTS index for the tile of interest (top-left tile)
+
+    Returns:
+        results_df (dataframe): results dataframe with additional columns that were calculated
+    """
+    results_df = results.pandas().xyxy[0] #:0 gets the dataframe, otherwise it would be a list
+    
+    #: check for empty dataframe
+    if results_df.empty:
+        logging.info('empty dataframe, no cooling towers detected')
+
+        return results_df
+    
+    zoom_level = 20
+
+    #: get upper-left coordinates of the primary tile (upper-left tile)
+    tile = mercantile.ul(int(col), int(row), zoom_level)
```

Comment:
**question**: what happens if col or row are not integers?

Reply:
it throws an error:
`TypeError: '<=' not supported between instances of 'int' and 'str'`

Reply:
For this one, we're already casting the col/row values as ints in the function call.  Only strings of integers will be read into the row/col variables from our BQ table.  So what else should be done to be more defensive?

Reply:
Generally speaking it's unsafe to assume that a database or user input will always be what you expect. This is the first time in this function the variables are cast to int. If it's not the the first time in the flow then the cast is unnecessary as you can ensure they are ints already. It's trivial to have a type check or a try except and exit early so you don't bomb out with an unhandled ValueError.